Generate "Golden Data" trace spans

kbrockhoff opened the pull request:
**Description:** This pull request has two items:
1. PICT tool input files for trace data and resulting output files from running pict from the command line. These are in `internal/goldendataset/testdata`. See: [Pairwise Independent Combinatorial Testing](https://github.com/microsoft/pict)
2. Generator for fully-populated OLTP ResourceSpans from PICT output files.

This has all the intended functionality for this PR. It has no impact on other functionality so there should be no problem in merging to master. I will follow up with other PRs to complete full "Golden Dataset" testing functionality.

**Link to tracking Issue:** #652 

**Testing:** Unit tests exist for all functionality which has been coded so far

**Documentation:** None yet. Will do so after golden dataset generation coding is complete.

tigrannajaryan commented on internal/goldendataset/span_generator.go:
```diff
@@ -0,0 +1,559 @@
+	}
+}
+
+func GenerateSpan(traceID []byte, tracestate string, parentID []byte, spanName string, kind string, spanTypeID string,
+	eventCnt string, linkCnt string, statusStr string) *otlptrace.Span {
```
Having so many parameters, especially when several are of the same type (string) can very easily lead to bugs due to incorrect ordering of parameters at call site. It is better to define a struct type with these parameters as fields and pass it to the func.
Alternatively, define type aliases for all these `string` types and their corresponding const declarations to help compiler catch misuse.

tigrannajaryan commented on internal/goldendataset/span_generator.go:
```diff
@@ -0,0 +1,559 @@
+
+func GenerateSpan(traceID []byte, tracestate string, parentID []byte, spanName string, kind string, spanTypeID string,
+	eventCnt string, linkCnt string, statusStr string) *otlptrace.Span {
+	endTime := time.Now().Add(-50 * time.Microsecond)
```
Nit: `time.Sub` may be slightly easier to understand than `time.Add` of negative duration.

tigrannajaryan commented on internal/goldendataset/span_generator.go:
```diff
@@ -0,0 +1,559 @@
+		parentID = spanList[i].SpanId
+		index++
+		if index >= pairsTotal {
+			index = 1
```
A comment would help to understand that we wrap around (and why 1 and not 0).

tigrannajaryan commented on internal/goldendataset/span_generator.go:
```diff
@@ -0,0 +1,559 @@
+
+func generateTraceID() []byte {
+	var r [16]byte
+	_, err := rand.Read(r[:])
```
Do we want the generator to be trully pseudo-random?
It may be very useful to instead pass an instance of rand.Rand to the generator. This will allow to pre-seed the random generator with the same value in comparison tests that use the generator. The seed may still be random for each execution of the test suite but it is important that comparison tests use the same seed and same sequence of random data so that comparison is valid.

This will require to make the generator a statefull object where Rand instance can live but I think it will make the generator more widely useful.

This is important for example for tests which can exhibit different results depending on the content of data, e.g. benchmarks that calculate compression ratio of different compression algorithms for the same input data.

kbrockhoff replied:
Changed so the generator is passed in from top-level generation method calls

tigrannajaryan commented on internal/goldendataset/span_generator.go:
```diff
@@ -0,0 +1,559 @@
+	ColumnStatus     = 6
+)
+
+const (
```
I think all these const sections will benefit it each had a corresponding type alias that declares a strong type to use in these consts and later in the code instead of using `string` for everything. With everything being `string` it is easy to make accidental mistakes.

kbrockhoff replied:
Added types

tigrannajaryan commented on internal/goldendataset/testdata/pict_pairs_spans.txt:
```diff
@@ -0,0 +1,289 @@
+Parent	Tracestate	Kind	Attributes	Events	Links	Status
```
Is this is generated file? Can we have `generated` in the name and document how it is generated/can be regenerated?

kbrockhoff replied:
Changed file names to start with generated. I plan to do documentation after the entire solution is coded and I have chance to play around with the best approach for making changes and automating it all.

tigrannajaryan commented on translator/conventions/opentelemetry.go:
```diff
@@ -15,18 +15,22 @@
-	AttributeLibraryName      = "library.name"
-	AttributeLibraryLanguage  = "library.language"
-	AttributeLibraryVersion   = "library.version"
+	AttributeLibraryName      = "telemetry.sdk.name"
+	AttributeLibraryLanguage  = "telemetry.sdk.language"
+	AttributeLibraryVersion   = "telemetry.sdk.version"
```
Please rename to include TelemetrySDK` in the name. I believe these are not used yet anywhere in the contrib so should be safe to rename.

kbrockhoff replied:
Fixed

tigrannajaryan commented on translator/conventions/opentelemetry.go:
```diff
@@ -15,18 +15,22 @@
 
 // OpenTelemetry Semantic Convention values for Resource attribute names.
-// See: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-resource-semantic-conventions.md
+// See: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/resource/semantic_conventions/README.md
```
Thanks for fixing.

kbrockhoff commented on internal/goldendataset/resource_generator.go:
```diff
@@ -0,0 +1,186 @@
+	index := 0
+	for key, value := range attrsMap {
+		valType := reflect.TypeOf(value)
+		if valType.Kind() == reflect.Int || valType.Kind() == reflect.Int64 {
```
Changed to same approach

kbrockhoff commented on internal/goldendataset/resource_generator_test.go:
```diff
@@ -0,0 +1,51 @@
+	for _, rscID := range resourceIds {
+		rsc := GenerateResource(rscID)
+		str := rsc.String()
+		t.Logf("Resource: %s", str)
```
Removed logging

kbrockhoff commented on internal/goldendataset/resource_generator.go:
```diff
@@ -0,0 +1,186 @@
+	index := 0
+	for key, value := range attrsMap {
+		valType := reflect.TypeOf(value)
+		if valType.Kind() == reflect.Int || valType.Kind() == reflect.Int64 {
```
Changed to switch

kbrockhoff commented on internal/goldendataset/resource_generator.go:
```diff
@@ -0,0 +1,186 @@
+	}
+	return &otlpresource.Resource{
+		Attributes:             convertMapToAttributeKeyValues(attrs),
+		DroppedAttributesCount: uint32(len(attrs) % 4),
```
Changed to count if for sure 0 in at least some cases

kbrockhoff commented on internal/goldendataset/resource_generator.go:
```diff
@@ -0,0 +1,186 @@
+		attrs = generateFassAttributes()
+	} else {
+		panic("invalid rscID")
+	}
```
Changed to switch

bogdandrutu commented on internal/goldendataset/resource_generator.go:
```diff
@@ -0,0 +1,142 @@
+import (
+	otlpresource "github.com/open-telemetry/opentelemetry-proto/gen/go/resource/v1"
+
+	"github.com/open-telemetry/opentelemetry-collector/translator/conventions"
```
You need to change the imports because we started to use vanityurl and now this looks like `go.opentelemetry.io/collector/translator/...`

kbrockhoff replied:
Merged latest from master and fixed

tigrannajaryan commented on internal/goldendataset/span_generator.go:
```diff
@@ -0,0 +1,497 @@
+	SpanStatusUnauthenticated:    "nstark is unknown user",
+}
+
+func GenerateSpans(count int, startPos int, pictFile string, random io.Reader) ([]*otlptrace.Span, int, error) {
```
Can we document int parameters and return types? It is non-obvious unless you read the function body.

kbrockhoff replied:
Documented all public functions

tigrannajaryan commented on internal/goldendataset/span_generator.go:
```diff
@@ -0,0 +1,497 @@
+			PICTInputStatus(inputs[SpansColumnStatus]), random)
+		parentID = spanList[i].SpanId
+		index++
+		if index >= pairsTotal {
```
Nit: the check probably should be at the start of the loop, not at the end, since in line 80 nothing prevents `index` from being more than len(pairsData).

kbrockhoff replied:
Fixed

tigrannajaryan commented:
@kbrockhoff please fix the conflicts.

kbrockhoff commented:
Made two changes:

1. Switched to struct holding input values so easier to expand in the future
2. Changed generated span name to a concatenation of input values to make it easier to identify which spans did not translate with equality
